### PR TITLE
Multi version

### PR DESCRIPTION
**Pull request by MSoegtropIMC**
This PR enables support for selecting a package list version. Currently supported are:
- 8.13
- 8.12
- dev

Reenabling the dev branch seems to have issues on Mac, so let's see what CI says ...

@gares : except for the issues with the dev branch, this is very close to final.

### CODE REVIEW

**Comment by Zimmi48** [comment]
If switch names support `+`, I would really recommend using it before the Coq version to distinguish it from the platform version.

**Comment by MSoegtropIMC** [comment]
> If switch names support +, I would really recommend using it before the Coq version to distinguish it from the platform version.

Yes, I thought that the . gets hard to read. I will try "+". I also thought about using "_", but I like "+".

**Comment by Zimmi48** [comment]
One issue with `+` than I didn't realize when commenting is that if we keep using it for beta versions there might be switches with two `+` signs that do not mean the same thing (`2021.07+beta1+8.13`). So we should either stop using it for beta versions (`2021.07beta1+8.13`) or use a different symbol for the Coq version (`2021.07+beta1_8.13`).

**Comment by MSoegtropIMC** [comment]
So how about "__coq-platform_2021.02.1_8.13" so using "_" also es separator between coq-platform and its version? Btw. It used to be "_coq-platform_.2021.02.1". I decided to start with two underscores as classical prefix for "system reserved names in a user name space" but don't have strong opinions about this.

**Comment by Zimmi48** [comment]
Yes, your proposal looks pretty great!

**Comment by gares** [comment]
Overall it looks ok, but since you have coq_packages.8.12.sh I guess I don't get exactly what the plan is. 2021.02.x was, in my mental model, pivoted on Coq 8.13.x and, at least for some packages, there is no choice (coq-elpi 1.9 does not work on 8.12, you need 1.8 for that). Maybe you want to have coq_package_2021.02.x.sh  and, say, coq_packages_2021.06.x.sh in the same repo, each pivoted on its coq version, but this PR fixed one platform version and multiple package lists.

Can you clarify this and how git branches would then be used?

**Comment by MSoegtropIMC** [comment]
@gares: For packages which did exist for older versions of Coq the idea is to add them to the respective package list - for packages which are new for Coq 8.X they are only added from 8.X onward. So the idea is that the package lists do increase also for old versions.

After pondering over this for a while I think that this is the easiest to maintain and most useful method for users.

I am not yet sure what the final branch structure will be. Most likely this will become the master branch and we release from the master branch since with a multi-version branch there is no good reason for doing releases from older Platform versions.

**Comment by gares** [comment]
I'm not so sure. What sould 2021.02.1 mean if by picking 8.12 or 8.13 the versions of the packages change?

**Comment by MSoegtropIMC** [comment]
> I'm not so sure I get it. What would 2021.02.1 mean if by picking 8.12 or 8.13 the versions of the packages change?

Well yes, one has to say `2021.02.1_8.12` or `2021.02.1_8.13` or `2021.02.1_dev` and probably would use `2021.02.1` as abbreviation for `2021.02.1_8.13`.

**Comment by gares** [comment]
Maybe It's just a detail, but I'd prefer `2021.02.1` (for 8.13) and `2021.02.1_8.12` for the "backport" thing (`2021.02.1~8.12` if made an opam package, so that it is `<` than the one without a coq version). If only because we can't change the version schema every day ;-)

**Comment by MSoegtropIMC** [comment]
It is an important detail and I agree that we should fix the naming now. IMO it is perfectly agreeable to say "The Coq Platform" is the one with the latest released Coq and then there are "compatibility versions" with newer (dev - maybe some day even a beta) or older Coq. So shall we go with that?

And indeed it does make sense to come up with a naming scheme, which would work as a name for opam packages as well - I didn't think about it.

I guess one would have a dev Coq Platform package in `extra-dev` but everything else should work naturally. So I guess if we use the short name for the latest release, the previous versions must be named e.g. "~8.12"? How about the dev version?

**Comment by Zimmi48** [comment]
>IMO it is perfectly agreeable to say "The Coq Platform" is the one with the latest released Coq and then there are "compatibility versions" with newer (dev - maybe some day even a beta) or older Coq. So shall we go with that?

:+1: 

>And indeed it does make sense to come up with a naming scheme, which would work as a name for opam packages as well - I didn't think about it.

I'm actually wondering if we will ever want to provide a meta opam package for the Coq platform. The scripts as they are, with their management of opam switches, are pretty convincing. In particular, the Coq platform pins the version of the underlying OCaml compiler. opam packages for the Coq platform would probably be worth the trouble only if we were to relax this constraint, right?

**Comment by MSoegtropIMC** [comment]
> I'm actually wondering if we will ever want to provide a meta opam package for the Coq platform. 

There are currently a few issues with this:
- I have patch opam packages required for Windows which I couldn't get accepted upstream. Likely the best fix for this would be to create a fixed gtk-sourceview package for cygwin. There might be other ways around this after introducing the local pkg-config concept in opam.
- I have .dev packages for OCaml stuff (menhir, gappa) and main opam doesn't have an extra-dev concept. I guess here the best solution would be to add them to Coq's extra-dev.
- The system sanity checks are helpful - on all systems. One would have to get some of this into opam or create a "check my system sanity" opam package.

Also I think it would help to get the Windows opam repo merged into the main opam repo - as is the patch repo is required from time to time, since without it merge delays are not manageable. And I don't think it would be good to change recommendations every other release.

So this is a goal, but not short term.

**Comment by MSoegtropIMC** [comment]
So the final conclusion is to name the switches:

- `__coq-platform.2021.02.1`
- `__coq-platform.2021.02.1~8.12`
- `__coq-platform.2021.02.1~dev`

provided this works. Note that I went back to have a dot between `coq-platform` and the version, because this is what one would use with `opam install` as well. So the package name would be the same except that the two underscores are removed. I would like to keep this, because as I said this are auto generated names in a user name space.

**Comment by Zimmi48** [comment]
>Also I think it would help to get the Windows opam repo merged into the main opam repo - as is the patch repo is required from time to time, since without it merge delays are not manageable. And I don't think it would be good to change recommendations every other release.

About this, have you noted #92?

**Comment by MSoegtropIMC** [comment]
@gares : I am looking at the snap file and wondering why it doesn't use the full coq platforms script. I guess it is required to install opam on its own, e.g. cause it needs to live in a specific folder, but one could still setup opam upfront and let the script setup the switch and the repos and do all the rest. Is there a good reason why you didn't do that?

**Comment by gares** [comment]
I think it does for the build process:
https://github.com/coq/platform/blob/2021.02/linux/snap/snapcraft.yaml.in#L23

It uses bits and pieces of it to generate the package description, IMO. I did that before realizing it is not updated automatically.
Is this what you are talking about?

**Comment by MSoegtropIMC** [comment]
@gares : thanks - O just didn't understand it. I will fix it then. I guess one can in a later PR change it such that it uses the same packaging logic as Windows and Mac, so that it would also work with manually extended opam switches.

**Comment by gares** [comment]
The best would be to:
- minimal sed on snapcraft.yml.in with no need to have opam & co around, just set the version string and the platform script flags
- from the build step of the snap
  - run the platform script
  - run a snap specific installation script (now inlined and not shared with win/mac, e.g. the blacklist of files etc)
  - print the package description for the publishing step (similar code to the one generating html for mac)
  
Just in case, we can't run the main platform script outside snap, and then copy the files inside, because it is snap which sets up gtk & co, and it has to stay that way, since installing the snap pulls in the gnome "platform" snap as a dependency.

The snap really contains an opam root, properly trimmed. We can take the files from there and putting them in a mirror layout, as we do in the other install scripts, if it comes handy to share more code with win/mac it is welcome, but if not there is really no need for it.

**Comment by gares** [comment]
let me merge some commits in hierarchy-builder/coq-master

File: opam/opam-coq-archive/extra-dev/packages/coq-hierarchy-builder/coq-hierarchy-builder.dev/opam [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+depends: [ "coq-elpi" { (>= "1.10.2" & < "1.11~") | = "dev" } ]
+conflicts: [ "coq-hierarchy-builder-shim" ]
+synopsis: "High level commands to declare and evolve a hierarchy based on packed classes"
+description: """
+Hierarchy Builder is a high level language to build hierarchies of algebraic structures and make these
+hierarchies evolve without breaking user code. The key concepts are the ones of factory, builder
+and abbreviation that let the hierarchy developer describe an actual interface for their library.
+Behind that interface the developer can provide appropriate code to ensure retro compatibility.
+"""
+tags: [ "logpath:HB" ]
```
Comment (gares) [inline opam/opam-coq-archive/extra-dev/packages/coq-hierarchy-builder/coq-hierarchy-builder.dev/opam:22]:
```suggestion
tags: [ "logpath:HB" ]
url {
      src: "git+https://github.com/math-comp/hierarchy-builder#coq-master"
}
```

Comment (gares) [reply opam/opam-coq-archive/extra-dev/packages/coq-hierarchy-builder/coq-hierarchy-builder.dev/opam:22]: This is the problem: if there is no URL stanza, then it probably fetches some old version which does not have the `build` make target (as the log report)

Comment (gares) [reply opam/opam-coq-archive/extra-dev/packages/coq-hierarchy-builder/coq-hierarchy-builder.dev/opam:22]: CC @MSoegtropIMC

**Comment by MSoegtropIMC** [comment]
> ~let me merge some commits in hierarchy-builder/coq-master~ No, this is not the problem. I believe the problem is the overlay which is incomplete.

Yes, I copied the opam file from hierarchy-builder master without much thinking. Thanks for the fix!

**Comment by MSoegtropIMC** [comment]
@gares : as discussed, I won't merge this into 2021.02 but instead push this as master and make this the new main branch then.

**Comment by gares** [comment]
OK, works for me

**Comment by gares** [comment]
maybe call this branch `main` (the new default on github)

**Comment by MSoegtropIMC** [comment]
Thanks for the hint - I will call it `main`!

**Comment by gares** [comment]
The snap is still disabled. What was the problem there?

**Comment by MSoegtropIMC** [comment]
It is not compatible with the multi version stuff. I fixed a few things (e.g. opam repo layout) but not all. E.g. one needs to be able to recover the name of the version file from the switch name and the like. So the names / interfaces of some scripts the snap packager uses have changed. It shouldn't be hard to fix (for me since I know what changed), but I didn't have time over the weekend.

**Comment by MSoegtropIMC** [comment]
This went into the main branch and won't be merged.

**Pull request closed without merging**